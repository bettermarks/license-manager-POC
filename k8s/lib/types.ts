import { ResourceRequirements } from "../imports/k8s";
/**
 * Stages
 */
export enum Stage {
  LOC = "loc",
  DEV = "dev",
  CI = "ci",
  PRO = "pro",
}

/**
 * Segments
 */
export enum Segment {
  LOC00 = "loc00",
  DEV00 = "dev00",
  CI00 = "ci00",
  CI01 = "ci01",
  PRO00 = "pro00",
}

/**
 * Defines the type defnition of labels in key-value format
 * - Applied to Nodepools when creating K8S cluster
 * - Applied to Deployment definitions at pod spec level when creating a deployment
 */
export type NodeSelector = {
  nodetype: string;
};

export type ApplicationConfig = {
  SEGMENT: string,
  LOG_FORMAT: string;
  LOG_LEVEL: LogLevel;
  APM_URL: string,
  APM_ENABLED: boolean,
  APM_TRANSACTION_SAMPLE_RATE: string,
  GLU_HIERARCHY_PROVIDER_URL: string,
};

/**
 * Deployment configuration
 */
export type DeploymentConfig = {
  /**
   * Migration job resources
   */
  migrationJobResources: ResourceRequirements;
  /**
   * Load fixtures job resources
   */
  loadFixturesJobResources: ResourceRequirements;
  /**
   * API resources
   */
  apiResources: ResourceRequirements;
  /**
   * API replicas
   */
  apiReplicas: number;
};

/**
 * Log level
 */
export enum LogLevel {
  INFO = "INFO",
  WARNING = "WARNING",
  ERROR = "ERROR",
  DEBUG = "DEBUG",
}

/**
 * Namespaces
 */
export enum Namespace {
  DEFAULT = "default",
  LICENSING = "licensing",
}
