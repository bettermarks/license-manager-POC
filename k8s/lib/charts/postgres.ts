import { Chart, ChartProps } from "cdk8s";
import {
  ConfigMap,
  ImagePullPolicy,
  ISecret,
  Secret,
  ServiceType,
} from "cdk8s-plus-24";
import { Construct } from "constructs";
import { IntOrString, KubeDeployment, KubeService } from "../../imports/k8s";

interface PostgresChartProps extends ChartProps {
  image: string;
  name: string;
}

export class PostgresChart extends Chart {
  /**
   * Secret with Postgres credentials
   *
   * keys:
   * - DATABASE_HOST
   * - DATABASE_PORT
   * - DATABASE_USER
   * - DATABASE_PASSWORD
   * - DATABASE_NAME
   * - POSTGRES_PASSWORD
   */
  readonly secret: ISecret;

  constructor(scope: Construct, id: string, props: PostgresChartProps) {
    super(scope, id, props);
    const { name, namespace } = props;
    const pgSecret = new Secret(this, "secret", {
      metadata: {
        name: name,
        namespace,
      },
      stringData: {
        DATABASE_PORT: "5432",
        DATABASE_USER: "postgres",
        DATABASE_PASSWORD: "postgres",
        DATABASE_NAME: "postgres",
        POSTGRES_PASSWORD: "postgres",
        POSTGRES_USER: "postgres",
        POSTGRES_DB: "postgres",
      },
    });

    const configMap = new ConfigMap(this, "initdb", {
      metadata: {
        name: `${name}-init-db-script`,
        namespace,
      },
      data: {}, // TODO: any init command?
    });

    new KubeDeployment(this, "pg-deployment", {
      metadata: {
        name,
        namespace,
        labels: {
          app: name,
        },
      },
      spec: {
        selector: {
          matchLabels: {
            app: name,
          },
        },
        replicas: 1,
        template: {
          metadata: {
            labels: {
              app: name,
            },
          },
          spec: {
            containers: [
              {
                name,
                image: props.image,
                imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
                ports: [{ containerPort: 5432 }],
                envFrom: [
                  {
                    secretRef: { name: pgSecret.name },
                  },
                ],
                readinessProbe: {
                  tcpSocket: {
                    port: IntOrString.fromNumber(5432),
                  },
                  initialDelaySeconds: 10,
                },
                livenessProbe: {
                  tcpSocket: {
                    port: IntOrString.fromNumber(5432),
                  },
                  initialDelaySeconds: 10,
                },
                volumeMounts: [
                  {
                    name: `${configMap.name}-volume`,
                    mountPath: "/docker-entrypoint-initdb.d/",
                  },
                ],
              },
            ],
            volumes: [
              {
                name: `${configMap.name}-volume`,
                configMap: {
                  name: configMap.name,
                },
              },
            ],
          },
        },
      },
    });

    const pgService = new KubeService(this, id, {
      metadata: {
        name,
        namespace,
      },
      spec: {
        type: ServiceType.CLUSTER_IP,
        ports: [
          {
            port: 5432,
          },
        ],
        selector: {
          app: name,
        },
      },
    });

    pgSecret.addStringData("DATABASE_HOST", pgService.name);
    // pgSecret.addStringData("DATABASE_PORT", "5432");
    this.secret = pgSecret;
  }
}
