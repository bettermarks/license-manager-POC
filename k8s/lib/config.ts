import { Quantity } from "../imports/k8s";

import {
  ApplicationConfig,
  DeploymentConfig,
  LogLevel,
  Segment,
} from "./types";

/**
 * Information to go in secrets
 *
 * DB related:
 * - DATABASE_HOST
 * - DATABASE_PORT
 * - DATABASE_USER
 * - DATABASE_PASSWORD
 * - DATABASE_NAME
 *
 * Application related:
 *
 */

/**
 * Information to go in config map
 */
export const APPLICATION_CONFIG: { [key: string]: ApplicationConfig } = {
  [Segment.LOC00]: {
    SEGMENT: Segment.LOC00,
    LOG_HANDLER: "console",
    LOG_LEVEL: LogLevel.DEBUG,
    DATABASE_USER: "postgres",
    DATABASE_HOST: "localhost",
    DATABASE_PORT: "5432",
    DATABASE_NAME: "licm",
    APM_URL: "https://apm.bettermarks.com",
    APM_ENABLED: false,
    APM_TRANSACTION_SAMPLE_RATE: "0.1",
  },
  [Segment.DEV00]: {
    SEGMENT: Segment.DEV00,
    LOG_HANDLER: "console",
    LOG_LEVEL: LogLevel.DEBUG,
    DATABASE_USER: "postgres",
    DATABASE_HOST: "localhost",
    DATABASE_PORT: "5432",
    DATABASE_NAME: "licm",
    APM_URL: "https://apm.bettermarks.com",
    APM_ENABLED: true,
    APM_TRANSACTION_SAMPLE_RATE: "0.1",
  },
  [Segment.CI00]: {
    SEGMENT: Segment.CI00,
    LOG_HANDLER: "console",
    LOG_LEVEL: LogLevel.DEBUG,
    DATABASE_USER: "postgres",
    DATABASE_HOST: "localhost",
    DATABASE_PORT: "5432",
    DATABASE_NAME: "licm",
    APM_URL: "https://apm.bettermarks.com",
    APM_ENABLED: true,
    APM_TRANSACTION_SAMPLE_RATE: "0.1",
  },
  [Segment.PRO00]: {
    SEGMENT: Segment.PRO00,
    LOG_HANDLER: "console",
    LOG_LEVEL: LogLevel.DEBUG,
    DATABASE_USER: "postgres",
    DATABASE_HOST: "localhost",
    DATABASE_PORT: "5432",
    DATABASE_NAME: "licm",
    APM_URL: "https://apm.bettermarks.com",
    APM_ENABLED: true,
    APM_TRANSACTION_SAMPLE_RATE: "0.1",
  },
};

/**
 * Deployment configuration per segment
 */
export const DEPLOYMENT_CONFIG: {
  [key: string]: DeploymentConfig;
} = {
  [Segment.DEV00]: {
    migrationJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.1),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    loadFixturesJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.1),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    apiResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("512Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("1024Mi"),
      },
    },
    apiReplicas: 3,
  },
  [Segment.CI00]: {
    migrationJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    loadFixturesJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    apiResources: {
      requests: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("512Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(2),
        memory: Quantity.fromString("1024Mi"),
      },
    },
    apiReplicas: 3,
  },
  [Segment.PRO00]: {
    migrationJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    loadFixturesJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    apiResources: {
      requests: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("512Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(2),
        memory: Quantity.fromString("1024Mi"),
      },
    },
    apiReplicas: 3,
  },
};
