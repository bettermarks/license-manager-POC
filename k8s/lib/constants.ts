import { NodeSelector } from "./types";

export const APP_NODE_POOL_LABELS: NodeSelector = {
  nodetype: "application",
};

/**
 * Name of the secret which is used by service chart to look up application secret.
 * This secret is created by AWS Secrets helper in bm-operations project. https://github.com/bettermarks/bm-operations/blob/master/cdk8s/apps/licensing.ts
 */
export const APPLICATION_SECRET = "licensing-secret";
/**
 * Name of the secret which is used by service chart to look up Postgres DB credentials.
 * This secret is created  by AWS Secrets helper in bm-operations project. https://github.com/bettermarks/bm-operations/blob/master/cdk8s/apps/licensing.ts
 */
export const POSTGRES_SECRET = "licensing-postgres-secret";
/**
 * Name of the secret used by Service Account to pull image from AWS ECR registry.
 * This secret is created by AWS Registry Helper Cron. https://github.com/bettermarks/bm-operations/blob/master/cdk8s/apps/licensing.ts
 */
export const REGISTRY_CREDENTIALS = "registry-credentials";
/**
 * Name of the image used for postgres database.
 */
export const POSTGRES_IMAGE = "postgres:14";
