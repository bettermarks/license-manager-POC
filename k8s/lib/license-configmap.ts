import { ConfigMap } from "cdk8s-plus-24";
import { Construct } from "constructs";
import { ApplicationConfig } from "./types";

export type LicenseConfigMapProps = {
  name: string;
  appConfig: ApplicationConfig;
  namespace?: string;
};

export class LicenseConfigMap extends Construct {
  readonly configMap: ConfigMap;

  constructor(scope: Construct, id: string, props: LicenseConfigMapProps) {
    super(scope, id);
    const { name, namespace, appConfig } = props;

    this.configMap = new ConfigMap(this, `${name}-configmap`, {
      metadata: {
        name: `${name}-configmap`,
        namespace,
      },
    });
    for (const [key, value] of Object.entries(appConfig)) {
      if (Array.isArray(value) || typeof value === "boolean") {
        this.configMap.addData(key, JSON.stringify(value));
      } else {
        this.configMap.addData(key, value);
      }
    }
  }
}