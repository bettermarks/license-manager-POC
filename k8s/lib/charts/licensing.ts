import { Chart, ChartProps } from "cdk8s";
import { HttpIngressPathType, ImagePullPolicy } from "cdk8s-plus-24";
import { Construct } from "constructs";
import {
  EnvFromSource,
  IntOrString,
  KubeIngress, ResourceRequirements
} from "../../imports/k8s";

import { APPLICATION_CONFIG } from "../config";
import { LicensingConfigMap } from "../licensing-configmap";
import { LicensingService } from "../services/licensing-service";
import { Namespace, NodeSelector, Segment } from "../types";

/**
 * This class is the implementation detail of Licensing deployment.
 */
export type LicensingChartProps = ChartProps & {
  /**
   * Docker image for Licensing
   */
  image: string;
  /**
   * Name of the secret containing Postgres credentials
   *
   * Required keys:
   * - DB_HOST
   * - DB_PORT
   * - DB_USER
   * - DB_PASSWORD
   * - DB_NAME
   */
  postgresSecret: string;
  /**
   * Name of the secret containing necessary credentials for running the service.
   *
   * Required keys:
   */
  applicationSecret?: string;
  serviceAccountName: string;
  segment: Segment;
  name: string;
  nodeSelector?: NodeSelector;
  apiResources?: ResourceRequirements;
  /**
   * API replicas
   * @default 3
   */
  apiReplicas?: number;
};

/**
 * Deploys Licensing server.
 */
export class LicensingChart extends Chart {
  constructor(scope: Construct, id: string, props: LicensingChartProps) {
    super(scope, id, props);

    const {
      namespace = Namespace.LICENSING,
      apiReplicas = 1,
      apiResources,
      applicationSecret,
      image,
      nodeSelector,
      postgresSecret,
      segment,
      serviceAccountName,
      name,
    } = props;

    const applicationEnv: EnvFromSource[] = [
      {
        configMapRef: {
          name: new LicensingConfigMap(this, `${name}-configmap`, {
            appConfig: APPLICATION_CONFIG[segment],
            name,
            namespace,
          }).configMap.name,
        },
      },
      {
        secretRef: {
          name: postgresSecret,
        },
      },
      ...(applicationSecret
        ? [
          {
            secretRef: {
              name: applicationSecret,
            },
          },
        ]
        : []),
    ];

    const apiName = `${name}-api`;
    /**
     * DeploymentID is a unique identifier for each deployment
     */
    const deploymentId = Date.now().toString(16); // TODO: this should be rethought
    const licensingService = new LicensingService(this, `${apiName}-service`, {
      name: apiName,
      namespace,
      replicas: apiReplicas,
      serviceAccountName,
      nodeSelector: nodeSelector,
      deploymentId: `${apiName}-${deploymentId}`,

      initContainers: [
      ],
      containers: [
        {
          name: apiName,
          image,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          ports: [
            {
              containerPort: 8000,
            },
          ],
          resources: apiResources,
          envFrom: applicationEnv,
          readinessProbe: {
            httpGet: {
              port: IntOrString.fromNumber(8000),
              path: "/v1/status",
            },
            initialDelaySeconds: 10,
          },
          livenessProbe: {
            httpGet: {
              port: IntOrString.fromNumber(8000),
              path: "/v1/status",
            },
            initialDelaySeconds: 10,
          },
        },
      ],
      servicePort: 80,
      containerPort: 8000,
    });

    if (segment === Segment.LOC00) {
      /**
       * Create ingress
       */
      new KubeIngress(this, `${name}-ingress`, {
        metadata: {
          name: `${name}-ingress`,
        },
        spec: {
          ingressClassName: "nginx",
          rules: [
            {
              http: {
                paths: [
                  {
                    path: "/",
                    pathType: HttpIngressPathType.PREFIX,
                    backend: {
                      service: {
                        name: licensingService.service!.name,
                        port: {
                          number: 80,
                        },
                      },
                    },
                  },
                ],
              },
            },
          ],
        },
      });
    }
  }
}
