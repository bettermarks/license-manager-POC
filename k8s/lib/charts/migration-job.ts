import { Chart, ChartProps } from "cdk8s";
import { ImagePullPolicy } from "cdk8s-plus-24";
import {
  EnvFromSource,
  Quantity,
  ResourceRequirements,
} from "../../imports/k8s";
import { Construct } from "constructs";

import { NodeSelector, Segment, Namespace } from "../types";
import { POSTGRES_IMAGE } from "../constants";
import { APPLICATION_CONFIG } from "../config";
import { LicensingConfigMap } from "../licensing-configmap";
import { LicensingJob } from "../jobs";


/**
 * This class is the implementation detail of Licensing deployment.
 */
export type LicensingChartProps = ChartProps & {
  /**
   * Docker image for Licensing
   */
  image: string;
  /**
   * Name of the secret containing Postgres credentials
   *
   * Required keys:
   * - DB_HOST
   * - DB_PORT
   * - DB_USER
   * - DB_PASSWORD
   * - DB_NAME
   */
  postgresSecret: string;
  /**
   * Name of the secret containing necessary credentials for running the service.
   *
   * Required keys:
   */
  applicationSecret?: string;
  segment: Segment;
  name: string;
  serviceAccountName: string;
  nodeSelector?: NodeSelector;
  migrationJobResources?: ResourceRequirements;
};

export class MigrationJobChart extends Chart {
  constructor(scope: Construct, id: string, props: LicensingChartProps) {
    super(scope, id, props);

    const {
      namespace = Namespace.LICENSING,
      applicationSecret,
      image,
      nodeSelector,
      migrationJobResources,
      serviceAccountName,
      postgresSecret,
      segment,
      name,
    } = props;

    // Ideally we should not create same config map again
    const applicationEnv: EnvFromSource[] = [
      {
        configMapRef: {
          name: new LicensingConfigMap(this, `${name}-configmap`, {
            appConfig: APPLICATION_CONFIG[segment],
            name,
            namespace,
          }).configMap.name,
        },
      },
      {
        secretRef: {
          name: postgresSecret,
        },
      },
      ...(applicationSecret
        ? [
          {
            secretRef: {
              name: applicationSecret,
            },
          },
        ]
        : []),
    ];
    /**
     * Job for running database migration
     */
    new LicensingJob(this, "migrate", {
      name: `${name}-migration`,
      namespace,
      nodeSelector: nodeSelector,
      serviceAccountName,
      ttlSecondsAfterFinished: 60,
      initContainers: [
        {
          name: "wait-for-database-migration",
          image: POSTGRES_IMAGE,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          command: ["sh", "-c", "until pg_isready --host ${DB_HOST}; do sleep 1; done"],
          envFrom: applicationEnv,
          resources: {
            requests: {
              cpu: Quantity.fromNumber(0.1),
              memory: Quantity.fromString("32Mi"),
            },
            limits: {
              cpu: Quantity.fromNumber(0.2),
              memory: Quantity.fromString("64Mi"),
            },
          },
        },
      ],
      containers: [
        {
          name: "licensing-migration",
          image,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          command: ["bash", "-c"],
          args: ["alembic upgrade head"],
          envFrom: applicationEnv,
          resources: migrationJobResources,
        },
      ],
    });
  }
}