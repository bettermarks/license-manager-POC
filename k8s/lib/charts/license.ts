import { Chart, ChartProps } from "cdk8s";
import { HttpIngressPathType, ImagePullPolicy } from "cdk8s-plus-24";
import {
  EnvFromSource,
  IntOrString,
  KubeIngress,
  Quantity,
  ResourceRequirements,
  KubeRole,
  KubeServiceAccount,
  KubeRoleBinding
} from "../../imports/k8s";
import { Construct } from "constructs";
import { NodeSelector, Segment } from "../types";

import { LicenseService } from "../services/license-service";
import { APPLICATION_CONFIG } from "../config";
import { POSTGRES_IMAGE } from "../constants";
import { LicenseConfigMap } from "../license-configmap";


/**
 * This class is the implementation detail of License deployment.
 */
export type LicenseChartProps = ChartProps & {
  /**
   * Docker image for License
   */
  image: string;
  /**
   * Name of the secret containing Postgres credentials
   *
   * Required keys:
   * - DATABASE_HOST
   * - DATABASE_PORT
   * - DATABASE_USER
   * - DATABASE_PASSWORD
   * - DATABASE_NAME
   */
  postgresSecret: string;
  /**
   * Name of the secret containing necessary credentials for running the service.
   *
   * Required keys:
   */
  applicationSecret?: string;
  segment: Segment;
  name: string;
  nodeSelector?: NodeSelector;
  migrationJobResources?: ResourceRequirements;
  loadFixturesJobResources?: ResourceRequirements;
  apiResources?: ResourceRequirements;
  /**
   * API replicas
   * @default 3
  */
  apiReplicas?: number;
  /**
   * Secret names for AWS ECR registry
   * @default []
  */
  imagePullSecrets?: ReadonlyArray<string>;
};

/**
 * Deploys License server.
 */
export class LicenseChart extends Chart {
  constructor(scope: Construct, id: string, props: LicenseChartProps) {
    super(scope, id, props);

    const {
      namespace,
      apiReplicas = 1,
      apiResources,
      applicationSecret,
      image,
      loadFixturesJobResources,
      nodeSelector,
      migrationJobResources,
      postgresSecret,
      segment,
      name,
      imagePullSecrets = []
    } = props;

    const serviceAccount = new KubeServiceAccount(this, `${name}-service-account`, {
      metadata: {
        name: `${name}-service-account`,
        namespace,
      },
      imagePullSecrets: imagePullSecrets?.map((secretRef) => ({ name: secretRef })),
    });

    const role = new KubeRole(this, `${name}-role`, {
      metadata: {
        name: `${name}-role`,
        namespace,
      },
      rules: [
        {
          apiGroups: [""],
          resources: ["pods"],
          verbs: ["get", "list", "watch"],
        },
      ],
    });

    new KubeRoleBinding(this, `${name}-role-binding`, {
      metadata: {
        name: `${name}-role-binding`,
        namespace,
      },
      roleRef: {
        apiGroup: role.apiGroup,
        kind: role.kind,
        name: role.name,
      },
      subjects: [{ kind: serviceAccount.kind, name: serviceAccount.name }],
    });

    const applicationEnv: EnvFromSource[] = [
      {
        configMapRef: {
          name: new LicenseConfigMap(this, `${name}-configmap`, {
            appConfig: APPLICATION_CONFIG[segment],
            name,
            namespace
          }).configMap.name,
        },
      },
      {
        secretRef: {
          name: postgresSecret,
        },
      },
      ...(applicationSecret)
        ? [
          {
            secretRef: {
              name: applicationSecret,
            },
          },
        ]
        : []
    ];

    const apiName = `${name}-api`;
    /**
     * DeploymentID is a unique identifier for each deployment
     */
    const deploymentId = Date.now().toString(16);  // TODO: this should be rethought
    const licenseService = new LicenseService(this, `${apiName}-service`, {
      name: apiName,
      namespace,
      replicas: apiReplicas,
      serviceAccountName: serviceAccount.name,
      nodeSelector: nodeSelector,
      deploymentId: `${apiName}-${deploymentId}`,

      initContainers: [
        {
          name: `${apiName}-wait-for-database`,
          image: POSTGRES_IMAGE,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          command: ["sh", "-c", "until pg_isready --host ${DATABASE_HOST}; do sleep 1; done"],
          envFrom: applicationEnv,
          resources: {
            requests: {
              cpu: Quantity.fromNumber(0.1),
              memory: Quantity.fromString("32Mi"),
            },
            limits: {
              cpu: Quantity.fromNumber(0.2),
              memory: Quantity.fromString("64Mi"),
            },
          },
        },
        {
          name: `${apiName}-migration`,
          image,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          command: ["bash", "-c"],
          args: [
            "alembic upgrade head"
          ],
          envFrom: applicationEnv,
          resources: migrationJobResources,
        },
        {
          name: `${apiName}-load-fixtures`,
          image,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          command: ["bash", "-c"],
          args: [
            `echo "TODO: load_initial_products here"`,
          ],
          envFrom: applicationEnv,
          resources: loadFixturesJobResources,
        },
      ],
      containers: [
        {
          name: apiName,
          image,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          ports: [
            {
              containerPort: 8000,
            },
          ],
          resources: apiResources,
          envFrom: applicationEnv,
          readinessProbe: {
            httpGet: {
              port: IntOrString.fromNumber(8000),
              path: "/v1/status",
            },
            initialDelaySeconds: 10,
          },
          livenessProbe: {
            httpGet: {
              port: IntOrString.fromNumber(8000),
              path: "/v1/status",
            },
            initialDelaySeconds: 10,
          },
        },
      ],
      servicePort: 80,
      containerPort: 8000,
    });

    /**
     * Create ingress
     */
    new KubeIngress(this, `${name}-ingress`, {
      metadata: {
        name: `${name}-ingress`,
      },
      spec: {
        ingressClassName: "nginx",
        rules: [
          {
            host: "license.bettermarks.loc",
            http: {
              paths: [
                {
                  path: "/",
                  pathType: HttpIngressPathType.PREFIX,
                  backend: {
                    service: {
                      name: licenseService.service!.name,
                      port: {
                        number: 80,
                      },
                    },
                  },
                },
              ],
            },
          },
        ],
      },
    });
  }
}
