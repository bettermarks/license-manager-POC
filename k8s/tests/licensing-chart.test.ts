import { Testing } from "cdk8s";
import { LicensingChart } from "../lib/charts/licensing";
import { DEPLOYMENT_CONFIG } from "../lib/config";
import { Segment } from "../lib/types";
import { APP_NODE_POOL_LABELS } from "../lib/constants";
const segment = Segment.DEV00;
describe("licensing-chart", () => {
  beforeEach(() => {
    jest.useFakeTimers("modern").setSystemTime(new Date("2008-12-01"));
  });
  test("should match with snapshot", () => {
    const app = Testing.app();
    const chart = new LicensingChart(app, "test-chart", {
      name: "licensing",
      namespace: "dummy-namespace",
      image:
        "676249682729.dkr.ecr.eu-central-1.amazonaws.com/bm-glu:e6b588df29edbf984d876e195bdaee5230c5ad92",
      postgresSecret: "postgres-secret",
      applicationSecret: "application-secret",
      serviceAccountName: "dummy-service-account",
      nodeSelector: APP_NODE_POOL_LABELS,
      segment: segment,
      apiResources: DEPLOYMENT_CONFIG[segment].apiResources,
      apiReplicas: DEPLOYMENT_CONFIG[segment].apiReplicas,
    });
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });
});
