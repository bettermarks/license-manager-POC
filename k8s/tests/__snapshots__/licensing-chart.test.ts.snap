// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`licensing-chart should match with snapshot 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "APM_ENABLED": "true",
      "APM_TRANSACTION_SAMPLE_RATE": "0.1",
      "APM_URL": "https://apm.bettermarks.com",
      "LOG_FORMAT": "json",
      "LOG_LEVEL": "DEBUG",
      "SEGMENT": "dev00",
    },
    "immutable": false,
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "licensing-configmap",
      "namespace": "dummy-namespace",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "licensing-api-deployment",
      "namespace": "dummy-namespace",
    },
    "spec": Object {
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "app": "licensing-api",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "licensing-api",
            "deploymentId": "licensing-api-11defd82800",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "envFrom": Array [
                Object {
                  "configMapRef": Object {
                    "name": "licensing-configmap",
                  },
                },
                Object {
                  "secretRef": Object {
                    "name": "postgres-secret",
                  },
                },
                Object {
                  "secretRef": Object {
                    "name": "application-secret",
                  },
                },
              ],
              "image": "676249682729.dkr.ecr.eu-central-1.amazonaws.com/bm-glu:e6b588df29edbf984d876e195bdaee5230c5ad92",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "httpGet": Object {
                  "path": "/v1/status",
                  "port": 8000,
                },
                "initialDelaySeconds": 10,
              },
              "name": "licensing-api",
              "ports": Array [
                Object {
                  "containerPort": 8000,
                },
              ],
              "readinessProbe": Object {
                "httpGet": Object {
                  "path": "/v1/status",
                  "port": 8000,
                },
                "initialDelaySeconds": 10,
              },
              "resources": Object {
                "limits": Object {
                  "cpu": 1,
                  "memory": "256Mi",
                },
                "requests": Object {
                  "cpu": 0.2,
                  "memory": "128Mi",
                },
              },
            },
          ],
          "initContainers": Array [],
          "nodeSelector": Object {
            "nodetype": "application",
          },
          "serviceAccountName": "dummy-service-account",
          "topologySpreadConstraints": Array [
            Object {
              "labelSelector": Object {
                "matchLabels": Object {
                  "app": "licensing-api",
                },
              },
              "maxSkew": 1,
              "topologyKey": "nodetype",
              "whenUnsatisfiable": "DoNotSchedule",
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "licensing-api",
      "namespace": "dummy-namespace",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 8000,
        },
      ],
      "selector": Object {
        "app": "licensing-api",
      },
      "type": "ClusterIP",
    },
  },
]
`;
